В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [1]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [2]:
import numpy as np
from statsmodels.stats.weightstats import ztest
from scipy import stats

Визначимо гіпотези для вибірки даних:

Н0 = середній ріст студентів 170

Н1 = середній ріст студентів значно > 170

In [3]:
null_mean_1 = 170
mean_heights = np.mean(heights)
sd_heights = np.std(heights)
sample_size_heights = len(heights)
alpha = 0.05

In [4]:
# перевірка результатів
print("Середнє вибірки: {:.2f}".format(mean_heights)) 
print("Стандартне відхилення вибірки: {:.2f}".format(sd_heights))

Середнє вибірки: 170.70
Стандартне відхилення вибірки: 4.65


In [5]:
ztest_Score_1, p_value_1 = ztest(heights, value = null_mean_1, alternative='larger')

In [6]:
ztest_Score_1, p_value_1

(0.8482235371587619, 0.19815674571091746)

In [7]:
if(p_value_1 <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


**Висновок**

Середній ріст випадкової вибірки = 170.7 

На рівні значущості 0.05 p-значення більше, тому немає достатніх доказів, щоб відхилити нульову гіпотезу про те, що середній ріст студентів не відрізняється від 170

також ми маємо значення z-score = 0.85, що показує нам, що нова вибірка не більша за середнє значення 170 з ймовірністю 80%

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [8]:
group_a_scores = [78.55, 72.25, 79.88, 75.00, 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.90, 79.04, 68.20, 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.70, 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.60,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.90, 79.11, 75.44, 73.50, 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.90, 73.29, 70.99, 74.19]
       

In [9]:
alpha = 0.05
mean_A = np.mean(group_a_scores)
mean_B = np.mean(group_b_scores)

Визначимо гіпотези для вибірки даних:

Н0 = середні оцінки групи A та групи B не відрізняються, тобто новий курс не ефективний (value=0)

Н1 = середні оцінки групи A та групи B відрізняються, тобто курс ефективний (value!=0)

In [10]:
mean_A, mean_B

(77.08125, 74.73875000000001)

виконаємо тест

In [11]:
ztest_Score_2, p_value_2 = ztest(group_a_scores, group_b_scores, value=0)
ztest_Score_2, p_value_2

(2.357354864751502, 0.01840564855252801)

In [12]:
if(p_value_2 <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


**Висновок**

Статистика для двох вибірок z-критерію становить 2.357, а відповідне p-value дорівнює 0.018

Оскільки p-value < 0.05, у нас є достатньо доказів, щоб відкинути нульову гіпотезу, про те, що середні оцінки групи A та групи B не відрізняються. 

Тому можна припустити, що курс є ефективним 

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть такий самий тест як в Завданні 2.1. Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

In [13]:
group_a_scores_temp = [78.55, 72.25, 79.88, 75.00]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.90, 79.11, 75.44, 73.50, 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.90, 73.29, 70.99, 74.19]

In [14]:
alpha = 0.05
mean_A_temp = np.mean(group_a_scores_temp)
mean_B = np.mean(group_b_scores)

In [15]:
mean_A_temp, mean_B

(76.42, 74.73875000000001)

виконаємо тест

In [16]:
ztest_Score_3, p_value_3 = ztest(group_a_scores_temp, group_b_scores, value=0)
ztest_Score_3, p_value_3

(0.7199058408597413, 0.47158297154365014)

In [18]:
if(p_value_3 <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


**Висновок**

Статистика для двох вибірок z-критерію зменшилась і становить 0.7199, а відповідне p-value збільшилось дорівнює 0.471

Порівняно з попереднім результатом ми маємо прийняти Н0, і вважати, що курс не ефективний.
В таких випадках необхідно використовувати інший метод, для більш правдивіших результатів, так як z-тест працює найкраще для великих вибірок

спробуємо використати t-тест метод для цього варіанту 

In [19]:
t_stat_32, p_value_32 = stats.ttest_ind(group_a_scores_temp, group_b_scores, equal_var=False)
t_stat_32, p_value_32 

(0.8981847990283107, 0.4186781031315412)

In [21]:
if(p_value_32 < alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


**Висновок**

Отже, p-значення дорівнює 0.41

Це більше за критичний рівень значущості 0.05, тому ми не маємо достатніх доказів, щоб відхилити нульову гіпотезу про те, що середні оцінки групи A та групи B не відрізняються.

Припускаю, що по даним результатим оцінок ще не можемо сказати чи курс буде ефективним. (також бачимо що це близькі результати з попереднім експерементом)


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [22]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

Найкраще використовувати парний t-тест (paired t-test). Цей тест порівнює середні значення двох пов'язаних груп для визначення, чи є різниця між ними статистично значущою.

In [23]:
alpha = 0.05
mean_before_training = np.mean(before_training)
mean_after_training = np.mean(after_training)

In [24]:
mean_before_training, mean_after_training

(50.885999999999996, 54.03600000000001)

In [25]:
t_stat_4, p_value_4 = stats.ttest_ind(before_training, after_training)
t_stat_4, p_value_4

(-1.3668871347639466, 0.17693568771368318)

In [26]:
if(p_value_4 <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


Отже, p-значення дорівнює 0.1769
Це значно менше за критичний рівень значущості 0.05, тому ми маємо достатні докази, щоб відхилити нульову гіпотезу. 
Це означає, що показники працівників значно покращилися після тренінгу.